File: fraudboost/tree.py
"""
Value-weighted decision tree for fraud detection gradient boosting.

Implements binary decision trees that split based on net savings maximization
rather than traditional information gain measures.
"""

import numpy as np
from typing import Optional, Tuple, Dict, Any
from dataclasses import dataclass


@dataclass
class SplitInfo:
    """Information about a tree split."""
    feature: int
    threshold: float
    gain: float
    left_value: float
    right_value: float
    n_samples_left: int
    n_samples_right: int
    missing_direction: int = -1  # -1 for left, 1 for right


class ValueWeightedDecisionTree:
    """
    Binary decision tree optimized for value-weighted fraud detection.
    
    Instead of maximizing information gain, splits are chosen to maximize
    net savings: (fraud_caught * amounts) - (false_positives * fp_cost)
    """
    
    def __init__(self,
                 max_depth: int = 6,
                 min_samples_leaf: int = 10,
                 min_samples_split: int = 20,
                 min_gain: float = 1e-7,
                 fp_cost: float = 100.0,
                 reg_lambda: float = 1.0):
        """
        Args:
            max_depth: Maximum tree depth
            min_samples_leaf: Minimum samples required in a leaf
            min_samples_split: Minimum samples required to split
            min_gain: Minimum gain required for a split
            fp_cost: Fixed cost of false positive (investigation cost)
            reg_lambda: L2 regularization strength for leaf values
        """
        self.max_depth = max_depth
        self.min_samples_leaf = min_samples_leaf
        self.min_samples_split = min_samples_split
        self.min_gain = min_gain
        self.fp_cost = fp_cost
        self.reg_lambda = reg_lambda
        
        # Tree structure (will be built during fit)
        self.tree_ = {}
        self.feature_importances_ = None
    
    def fit(self, X: np.ndarray, gradients: np.ndarray, hessians: np.ndarray,
            amounts: Optional[np.ndarray] = None) -> 'ValueWeightedDecisionTree':
        """
        Fit the decision tree using gradients and hessians from boosting.
        
        Args:
            X: Feature matrix (n_samples, n_features)
            gradients: Gradient values for each sample
            hessians: Hessian values for each sample
            amounts: Transaction amounts for value weighting
        """
        n_samples, n_features = X.shape
        
        if amounts is None:
            amounts = np.ones(n_samples)
        
        # Initialize feature importances
        self.feature_importances_ = np.zeros(n_features)
        
        # Build tree recursively
        self.tree_ = self._build_tree(X, gradients, hessians, amounts, depth=0)
        
        # Normalize feature importances
        if self.feature_importances_.sum() > 0:
            self.feature_importances_ /= self.feature_importances_.sum()
        
        return self
    
    def predict(self, X: np.ndarray) -> np.ndarray:
        """Predict leaf values for input samples."""
        n_samples = X.shape[0]
        predictions = np.zeros(n_samples)
        
        for i in range(n_samples):
            predictions[i] = self._predict_sample(X[i], self.tree_)
        
        return predictions
    
    def _build_tree(self, X: np.ndarray, gradients: np.ndarray, hessians: np.ndarray,
                    amounts: np.ndarray, depth: int) -> Dict[str, Any]:
        """Recursively build the decision tree."""
        n_samples, n_features = X.shape
        
        # Compute leaf value using Newton-Raphson step with regularization
        leaf_value = -np.sum(gradients) / (np.sum(hessians) + self.reg_lambda)
        
        # Base cases for stopping
        if (depth >= self.max_depth or 
            n_samples < self.min_samples_split or
            n_samples <= self.min_samples_leaf):
            return {
                'leaf': True,
                'value': leaf_value,
                'n_samples': n_samples
            }
        
        # Find best split
        best_split = self._find_best_split(X, gradients, hessians, amounts)
        
        if best_split is None or best_split.gain < self.min_gain:
            return {
                'leaf': True,
                'value': leaf_value,
                'n_samples': n_samples
            }
        
        # Update feature importance
        self.feature_importances_[best_split.feature] += best_split.gain
        
        # Split data
        left_mask, right_mask = self._split_data(X, best_split)
        
        # Recursively build left and right subtrees
        left_tree = self._build_tree(
            X[left_mask], gradients[left_mask], hessians[left_mask], 
            amounts[left_mask], depth + 1
        )
        
        right_tree = self._build_tree(
            X[right_mask], gradients[right_mask], hessians[right_mask], 
            amounts[right_mask], depth + 1
        )
        
        return {
            'leaf': False,
            'feature': best_split.feature,
            'threshold': best_split.threshold,
            'missing_direction': best_split.missing_direction,
            'left': left_tree,
            'right': right_tree,
            'n_samples': n_samples,
            'gain': best_split.gain
        }
    
    def _find_best_split(self, X: np.ndarray, gradients: np.ndarray, 
                        hessians: np.ndarray, amounts: np.ndarray) -> Optional[SplitInfo]:
        """Find the split that maximizes net savings gain."""
        n_samples, n_features = X.shape
        best_split = None
        best_gain = -np.inf
        
        # Current node statistics
        current_grad_sum = np.sum(gradients)
        current_hess_sum = np.sum(hessians)
        current_value = -current_grad_sum / (current_hess_sum + self.reg_lambda)
        
        for feature in range(n_features):
            feature_values = X[:, feature]
            
            # Handle missing values by trying both directions
            valid_mask = ~np.isnan(feature_values)
            if not np.any(valid_mask):
                continue
                
            valid_indices = np.where(valid_mask)[0]
            valid_values = feature_values[valid_indices]
            
            # Get unique thresholds to try
            unique_values = np.unique(valid_values)
            if len(unique_values) < 2:
                continue
            
            # Try thresholds between unique values
            thresholds = (unique_values[:-1] + unique_values[1:]) / 2
            
            for threshold in thresholds:
                # Split based on threshold
                left_mask_valid = valid_values <= threshold
                right_mask_valid = ~left_mask_valid
                
                left_indices = valid_indices[left_mask_valid]
                right_indices = valid_indices[right_mask_valid]
                
                if len(left_indices) < self.min_samples_leaf or len(right_indices) < self.min_samples_leaf:
                    continue
                
                # Compute gain for this split
                gain = self._compute_split_gain(
                    gradients[left_indices], hessians[left_indices],
                    gradients[right_indices], hessians[right_indices],
                    current_grad_sum, current_hess_sum
                )
                
                if gain > best_gain:
                    best_gain = gain
                    
                    left_value = -np.sum(gradients[left_indices]) / (np.sum(hessians[left_indices]) + self.reg_lambda)
                    right_value = -np.sum(gradients[right_indices]) / (np.sum(hessians[right_indices]) + self.reg_lambda)
                    
                    # Determine missing direction based on which side has more samples
                    missing_direction = -1 if len(left_indices) > len(right_indices) else 1
                    
                    best_split = SplitInfo(
                        feature=feature,
                        threshold=threshold,
                        gain=gain,
                        left_value=left_value,
                        right_value=right_value,
                        n_samples_left=len(left_indices),
                        n_samples_right=len(right_indices),
                        missing_direction=missing_direction
                    )
        
        return best_split
    
    def _compute_split_gain(self, left_grad: np.ndarray, left_hess: np.ndarray,
                           right_grad: np.ndarray, right_hess: np.ndarray,
                           total_grad: float, total_hess: float) -> float:
        """
        Compute the gain from a split using gradient boosting criterion.
        
        Gain = 0.5 * [L²/(H+λ) + R²/(H+λ) - (L+R)²/(H+λ)]
        where L,R are gradient sums and H is hessian sum.
        """
        left_grad_sum = np.sum(left_grad)
        left_hess_sum = np.sum(left_hess)
        right_grad_sum = np.sum(right_grad)
        right_hess_sum = np.sum(right_hess)
        
        left_score = (left_grad_sum ** 2) / (left_hess_sum + self.reg_lambda)
        right_score = (right_grad_sum ** 2) / (right_hess_sum + self.reg_lambda)
        parent_score = (total_grad ** 2) / (total_hess + self.reg_lambda)
        
        gain = 0.5 * (left_score + right_score - parent_score)
        
        return gain
    
    def _split_data(self, X: np.ndarray, split: SplitInfo) -> Tuple[np.ndarray, np.ndarray]:
        """Split data based on the split info."""
        feature_values = X[:, split.feature]
        
        # Handle missing values
        valid_mask = ~np.isnan(feature_values)
        missing_mask = np.isnan(feature_values)
        
        left_mask = np.zeros(len(X), dtype=bool)
        right_mask = np.zeros(len(X), dtype=bool)
        
        # Valid values
        left_mask[valid_mask] = feature_values[valid_mask] <= split.threshold
        right_mask[valid_mask] = feature_values[valid_mask] > split.threshold
        
        # Missing values go to the side with more samples
        if split.missing_direction == -1:  # Left
            left_mask[missing_mask] = True
        else:  # Right
            right_mask[missing_mask] = True
        
        return left_mask, right_mask
    
    def _predict_sample(self, x: np.ndarray, tree: Dict[str, Any]) -> float:
        """Predict a single sample by traversing the tree."""
        if tree['leaf']:
            return tree['value']
        
        feature_value = x[tree['feature']]
        
        # Handle missing values
        if np.isnan(feature_value):
            if tree['missing_direction'] == -1:
                return self._predict_sample(x, tree['left'])
            else:
                return self._predict_sample(x, tree['right'])
        
        # Normal split
        if feature_value <= tree['threshold']:
            return self._predict_sample(x, tree['left'])
        else:
            return self._predict_sample(x, tree['right'])
    
    def get_leaf_values(self) -> np.ndarray:
        """Get all leaf values from the tree."""
        leaf_values = []
        self._collect_leaf_values(self.tree_, leaf_values)
        return np.array(leaf_values)
    
    def _collect_leaf_values(self, tree: Dict[str, Any], leaf_values: list):
        """Recursively collect leaf values."""
        if tree['leaf']:
            leaf_values.append(tree['value'])
        else:
            self._collect_leaf_values(tree['left'], leaf_values)
            self._collect_leaf_values(tree['right'], leaf_values)

File: rust_core/src/tree.rs
use ndarray::{Array1, ArrayView1, ArrayView2};
use rayon::prelude::*;

#[derive(Debug, Clone)]
pub struct SplitInfo {
    pub feature: usize,
    pub threshold: f64,
    pub gain: f64,
    pub n_samples_left: usize,
    pub n_samples_right: usize,
}

#[derive(Debug, Clone)]
pub struct TreeNode {
    pub is_leaf: bool,
    pub value: f64,
    pub feature: Option<usize>,
    pub threshold: Option<f64>,
    pub left: Option<Box<TreeNode>>,
    pub right: Option<Box<TreeNode>>,
    pub n_samples: usize,
}

impl TreeNode {
    pub fn new_leaf(value: f64, n_samples: usize) -> Self {
        TreeNode {
            is_leaf: true,
            value,
            feature: None,
            threshold: None,
            left: None,
            right: None,
            n_samples,
        }
    }

    pub fn new_internal(
        feature: usize,
        threshold: f64,
        left: TreeNode,
        right: TreeNode,
        n_samples: usize,
    ) -> Self {
        TreeNode {
            is_leaf: false,
            value: 0.0, // Not used for internal nodes
            feature: Some(feature),
            threshold: Some(threshold),
            left: Some(Box::new(left)),
            right: Some(Box::new(right)),
            n_samples,
        }
    }

    pub fn predict(&self, x: &ArrayView1<f64>) -> f64 {
        if self.is_leaf {
            return self.value;
        }

        let feature_idx = self.feature.unwrap();
        let threshold = self.threshold.unwrap();
        let feature_value = x[feature_idx];

        // Handle missing values (NaN) by going to the larger child
        if feature_value.is_nan() {
            let left_samples = self.left.as_ref().unwrap().n_samples;
            let right_samples = self.right.as_ref().unwrap().n_samples;
            if left_samples >= right_samples {
                return self.left.as_ref().unwrap().predict(x);
            } else {
                return self.right.as_ref().unwrap().predict(x);
            }
        }

        if feature_value <= threshold {
            self.left.as_ref().unwrap().predict(x)
        } else {
            self.right.as_ref().unwrap().predict(x)
        }
    }
}

pub struct TreeBuilder {
    max_depth: usize,
    min_samples_leaf: usize,
    min_samples_split: usize,
    reg_lambda: f64,
    min_gain: f64,
}

impl TreeBuilder {
    pub fn new(
        max_depth: usize,
        min_samples_leaf: usize,
        min_samples_split: usize,
        reg_lambda: f64,
    ) -> Self {
        TreeBuilder {
            max_depth,
            min_samples_leaf,
            min_samples_split,
            reg_lambda,
            min_gain: 1e-7,
        }
    }

    pub fn build_tree(
        &self,
        x: &ArrayView2<f64>,
        gradients: &ArrayView1<f64>,
        hessians: &ArrayView1<f64>,
        sample_indices: &[usize],
        feature_indices: &[usize],
        depth: usize,
    ) -> TreeNode {
        let n_samples = sample_indices.len();

        // Compute leaf value using Newton-Raphson step
        let grad_sum: f64 = sample_indices.iter().map(|&i| gradients[i]).sum();
        let hess_sum: f64 = sample_indices.iter().map(|&i| hessians[i]).sum();
        let leaf_value = -grad_sum / (hess_sum + self.reg_lambda);

        // Check stopping conditions
        if depth >= self.max_depth
            || n_samples < self.min_samples_split
            || n_samples <= self.min_samples_leaf
        {
            return TreeNode::new_leaf(leaf_value, n_samples);
        }

        // Find the best split
        let best_split = self.find_best_split(
            x,
            gradients,
            hessians,
            sample_indices,
            feature_indices,
            grad_sum,
            hess_sum,
        );

        if let Some(split) = best_split {
            if split.gain >= self.min_gain
                && split.n_samples_left >= self.min_samples_leaf
                && split.n_samples_right >= self.min_samples_leaf
            {
                // Create the split
                let (left_indices, right_indices) =
                    self.split_samples(x, sample_indices, split.feature, split.threshold);

                let left_child = self.build_tree(
                    x,
                    gradients,
                    hessians,
                    &left_indices,
                    feature_indices,
                    depth + 1,
                );

                let right_child = self.build_tree(
                    x,
                    gradients,
                    hessians,
                    &right_indices,
                    feature_indices,
                    depth + 1,
                );

                return TreeNode::new_internal(
                    split.feature,
                    split.threshold,
                    left_child,
                    right_child,
                    n_samples,
                );
            }
        }

        // No valid split found, return leaf
        TreeNode::new_leaf(leaf_value, n_samples)
    }

    fn find_best_split(
        &self,
        x: &ArrayView2<f64>,
        gradients: &ArrayView1<f64>,
        hessians: &ArrayView1<f64>,
        sample_indices: &[usize],
        feature_indices: &[usize],
        total_grad: f64,
        total_hess: f64,
    ) -> Option<SplitInfo> {
        // Use rayon for parallel feature processing
        let splits: Vec<Option<SplitInfo>> = feature_indices
            .par_iter()
            .map(|&feature_idx| {
                self.find_best_split_for_feature(
                    x,
                    gradients,
                    hessians,
                    sample_indices,
                    feature_idx,
                    total_grad,
                    total_hess,
                )
            })
            .collect();

        // Find the best split across all features
        splits
            .into_iter()
            .filter_map(|s| s)
            .max_by(|a, b| a.gain.partial_cmp(&b.gain).unwrap_or(std::cmp::Ordering::Equal))
    }

    fn find_best_split_for_feature(
        &self,
        x: &ArrayView2<f64>,
        gradients: &ArrayView1<f64>,
        hessians: &ArrayView1<f64>,
        sample_indices: &[usize],
        feature_idx: usize,
        total_grad: f64,
        total_hess: f64,
    ) -> Option<SplitInfo> {
        // Collect feature values with their corresponding gradient/hessian
        let mut samples: Vec<(f64, f64, f64, usize)> = sample_indices
            .iter()
            .map(|&i| (x[[i, feature_idx]], gradients[i], hessians[i], i))
            .filter(|(val, _, _, _)| !val.is_nan()) // Filter out NaN values
            .collect();

        if samples.len() < 2 * self.min_samples_leaf {
            return None;
        }

        // Sort by feature value
        samples.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(std::cmp::Ordering::Equal));

        let mut best_split: Option<SplitInfo> = None;
        let mut best_gain = f64::NEG_INFINITY;

        let mut left_grad = 0.0;
        let mut left_hess = 0.0;

        // Try all possible thresholds (between unique values)
        for i in 0..samples.len() - 1 {
            left_grad += samples[i].1;
            left_hess += samples[i].2;

            let right_grad = total_grad - left_grad;
            let right_hess = total_hess - left_hess;

            // Only consider splits that satisfy min_samples_leaf
            let left_count = i + 1;
            let right_count = samples.len() - left_count;

            if left_count < self.min_samples_leaf || right_count < self.min_samples_leaf {
                continue;
            }

            // Use threshold between consecutive different values
            if samples[i].0 >= samples[i + 1].0 - f64::EPSILON {
                continue; // Skip if values are the same
            }

            let threshold = (samples[i].0 + samples[i + 1].0) / 2.0;

            // Compute gain using gradient boosting criterion
            let gain = self.compute_split_gain(left_grad, left_hess, right_grad, right_hess, total_grad, total_hess);

            if gain > best_gain {
                best_gain = gain;
                best_split = Some(SplitInfo {
                    feature: feature_idx,
                    threshold,
                    gain,
                    n_samples_left: left_count,
                    n_samples_right: right_count,
                });
            }
        }

        best_split
    }

    fn compute_split_gain(
        &self,
        left_grad: f64,
        left_hess: f64,
        right_grad: f64,
        right_hess: f64,
        total_grad: f64,
        total_hess: f64,
    ) -> f64 {
        // XGBoost/LightGBM gain formula:
        // Gain = 0.5 * [L²/(H+λ) + R²/(H+λ) - (L+R)²/(H+λ)]
        let left_score = (left_grad * left_grad) / (left_hess + self.reg_lambda);
        let right_score = (right_grad * right_grad) / (right_hess + self.reg_lambda);
        let parent_score = (total_grad * total_grad) / (total_hess + self.reg_lambda);

        0.5 * (left_score + right_score - parent_score)
    }

    fn split_samples(
        &self,
        x: &ArrayView2<f64>,
        sample_indices: &[usize],
        feature_idx: usize,
        threshold: f64,
    ) -> (Vec<usize>, Vec<usize>) {
        let mut left_indices = Vec::new();
        let mut right_indices = Vec::new();

        for &sample_idx in sample_indices {
            let feature_value = x[[sample_idx, feature_idx]];
            
            if feature_value.is_nan() {
                // For missing values, use the same logic as prediction (go to larger child)
                // This is a heuristic - we'll assign based on which side currently has more samples
                if left_indices.len() >= right_indices.len() {
                    left_indices.push(sample_idx);
                } else {
                    right_indices.push(sample_idx);
                }
            } else if feature_value <= threshold {
                left_indices.push(sample_idx);
            } else {
                right_indices.push(sample_idx);
            }
        }

        (left_indices, right_indices)
    }
}

/// Compute feature importances by summing gains
pub fn compute_feature_importances(tree: &TreeNode, n_features: usize) -> Array1<f64> {
    let mut importances = Array1::zeros(n_features);
    collect_importances(tree, &mut importances);
    
    // Normalize
    let total: f64 = importances.sum();
    if total > 0.0 {
        importances /= total;
    }
    
    importances
}

fn collect_importances(node: &TreeNode, importances: &mut Array1<f64>) {
    if !node.is_leaf {
        if let Some(feature) = node.feature {
            // In a real implementation, you'd store the gain in the node
            // For now, we'll use a simple heuristic based on samples
            importances[feature] += node.n_samples as f64;
            
            if let Some(ref left) = node.left {
                collect_importances(left, importances);
            }
            if let Some(ref right) = node.right {
                collect_importances(right, importances);
            }
        }
    }
}